OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs

```diff
@@ -45,39 +43,35 @@ public ClassicProductionQueue(ActorInitializer init, ClassicProductionQueueInfo
 			this.info = info;
 		}
 
-		[Sync] bool isActive = false;
-
-		public override void Tick(Actor self)
+		protected override void UpdateProductionTraitPairs()
 		{
-			// PERF: Avoid LINQ.
-			isActive = false;
-			foreach (var x in self.World.ActorsWithTrait<Production>())
-			{
-				if (x.Actor.Owner == self.Owner && x.Trait.Info.Produces.Contains(Info.Type))
-				{
-					isActive = true;
-					break;
-				}
-			}
+			allProductionTraitPairs = self.World.ActorsWithTrait<Production>().Where(x => x.Actor.Owner == self.Owner).ToArray();
+			allCompatibleProductionTraitPairs = allProductionTraitPairs
+				.Where(x => x.Trait.Info.Produces.Contains(Info.Type))
+				.ToArray();
+		}
 
-			base.Tick(self);
+		void OnActorAddedToOrRemovedFromWorld(Actor actor)
+		{
+			if (actor.Owner == self.Owner)
```

Comment:
You can check for `HasTrait<Production>()` here and avoid a lot of unnecessary recalculations.